### PR TITLE
ETCD-514: Add etcd size tuning

### PR DESCRIPTION
**Pull request by dusk125**
This enhancement describes allowing customers to tweak etcd backend quota in a controlled manor.

### CODE REVIEW

**Comment by dusk125** [comment]
ptal
@hasbro7 @tjungblu @Elbehery @williamcaban @soltysh

File: enhancements/etcd/etcd-size-tuning.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/ETCD-514
+---
+
+# etcd Size Tuning
+
+## Summary
+
+Thie enhancement would replace the hardcoded value for the etcd backend database quota with predefined profiles.
+Each profile would map to predefined, and pretested, values for the internal parameter.
```
[the remark this answers is not in the document]
Comment (hasbro17) [reply enhancements/etcd/etcd-size-tuning.md:26]: I like the idea of a quantity over enums too. Seems like we have precedence for that in the API already. Without any validation though.
https://github.com/openshift/api/blob/e2ce40831f2ec58c3aa06e0ecd1c17a647115c94/machineconfiguration/v1/types.go#L678-L687

I see resource.Quantity also allows milli values for specifying cores though.
https://github.com/kubernetes/apimachinery/blob/f14778da5523847e4c07346e3161a4b4f6c9186e/pkg/api/resource/quantity_example_test.go#L48

Beyond parsing and validation in the controller, would we need CEL validation to disallow that?

File: enhancements/etcd/etcd-size-tuning.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+#### Variation [optional]
+
+1. The cluster administrator wants to set the quota from a larger value (e.g. 32GiB) to a smaller one (8GiB).
+2. They attempt to set the new value in the API server.
+3. The API server should return an error notifying them that the value is invalid - not allowed to shrink the quota.
+
+### API Extensions
+
+None
```
Comment (hasbro17) [inline enhancements/etcd/etcd-size-tuning.md:77]: Can we outline the API here to discuss the naming/placement and validation.
i.e https://github.com/openshift/api/pull/1736

File: enhancements/etcd/etcd-size-tuning.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+* Handle consuming value changes without an etcd rollout.
+
+## Proposal
+
+Allow the customer/admin access to set the backend quota directly through a human readable value. The only restriction is that they are disallowed from setting the value lower than the current.
+
+The default value or unset value would be the current, hardcoded backend quota: 8GiB. This would allow for upgrades where this configuration value wouldn't nesscessarily be set.
+
+This would allow for more flexibility, but would also likely increase the number of support cases as the memory requirements will be higher: changing quota from 8GiB to 16GiB is not double the memory, it'll be 6x the memory since each etcd member could take up to 16GiB.
+We will need to work with the perf team to understand the memory requirements of largers quotas, and the effect of the quota on things like defragmentation.
```
Comment (hasbro17) [inline enhancements/etcd/etcd-size-tuning.md:59]: Didn't see an upper limit specified elsewhere so as part of the proposal we should specify an upper limit e.g 64 GB. 

With a note on the need to test with the perf scale team what a reasonable and safe upper bound would be before we GA this.

Also mention the upper limit in the API extensions section below so it's clear.

**Review by hasbro17 — COMMENTED** [review]
Gets the general idea across but need to mention some specifics on the API. This should streamline the actual API PR if we get those details outlined here first.

File: enhancements/etcd/etcd-size-tuning.md (hunk 3 above)
Comment (dhellmann) [reply enhancements/etcd/etcd-size-tuning.md:59]: Are there any other constraints on what values make sense? For example, does the limit need to be a power of 2? Is 8 the minimum?

File: enhancements/etcd/etcd-size-tuning.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+1. The cluster administrator wants to set the quota from a larger value (e.g. 32GiB) to a smaller one (8GiB).
+2. They attempt to set the new value in the API server.
+3. The API server should return an error notifying them that the value is invalid - not allowed to shrink the quota.
+
+### API Extensions
+
+In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
+- Add BackendSizeQuota of type *resource.Quantity
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply enhancements/etcd/etcd-size-tuning.md:80]: If the max is 64GiB and the minimum is 8GiB, and if CEL supports saying "this value may not shrink", then it might be more natural for the API field to have a name like `backendQuotaGiB` and be an integer.

How will an admin who changes the quota know when their change is done taking effect? Is there already a field in the status portion of the API to report the size being used, for example?

File: enhancements/etcd/etcd-size-tuning.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+N/A
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+N/A
```
Comment (dhellmann) [inline enhancements/etcd/etcd-size-tuning.md:156]: What happens if they ask for more space than is available, especially in an on-prem configuration where dynamic storage allocation may not be possible?

File: enhancements/etcd/etcd-size-tuning.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+#### Support Procedures
+
+- If the user attempts to set the quota to an invalid value (not in the form of a resource quantity (e.g. 8GiB, etc)), the API will not accept the value and return an error.
+- If the user attempts to set the quota to a lower value than the current (e.g. 21GiB -> 16GiB), the API will not accept the value and return an error.
```
Comment (dhellmann) [inline enhancements/etcd/etcd-size-tuning.md:161]: How can an admin monitor the rollout of the change? What if that rollout gets stuck (say 1 of the control plane hosts doesn't have enough space, but the other 2 do), where do they look for debugging information to understand what's going on?

File: enhancements/etcd/etcd-size-tuning.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+## Implementation History
+
+None
+
+## Alternatives
+
+### Discrete Profile values
```
Comment (dhellmann) [inline enhancements/etcd/etcd-size-tuning.md:169]: This seems like an interesting idea, why was it rejected?

File: enhancements/etcd/etcd-size-tuning.md (hunk 3 above)
Comment (hasbro17) [reply enhancements/etcd/etcd-size-tuning.md:59]: > does the limit need to be a power of 2

No limitation from etcd that I'm aware of. This limit just dictates the size of the DB file which is mapped into memory.

> Is 8 the minimum?

The upstream limit by default is 2GB https://etcd.io/docs/v3.5/dev-guide/limit/#storage-size-limit
However we've always configured this as 8GB by default for openshift.
We're not aiming to change the default yet, and we don't want to allow decreasing the limit so customers don't write lock the cluster if they accidentally reduce it to less than what's being used by their cluster.

For the upper limit, 64GB is a ballpark figure. Others have safely gone as high as 100GB.
https://www.cncf.io/blog/2019/05/09/performance-optimization-of-etcd-in-web-scale-data-scenario/

But there may be other constraints on the size of the DB file such as the time to restore, and transporting the snapshot to a new member could be slow depending on the hardware.

We'll have to test out those constraints on the upper limit that we choose here.

File: enhancements/etcd/etcd-size-tuning.md (hunk 5 above)
Comment (hasbro17) [reply enhancements/etcd/etcd-size-tuning.md:156]: That's a good point.
@dusk125 @Elbehery We should point this out by checking this failure mode and see if etcd has any mechanisms to detect this. I suspect no, and it will eventually exhaust the disk space if the limit is greater than the available storage and not applicable but would be good to confirm.

Or actually if the db in usage size grows larger than the available memory, I would think the etcd pod could get OOM killed as well.

File: enhancements/etcd/etcd-size-tuning.md (hunk 7 above)
Comment (hasbro17) [reply enhancements/etcd/etcd-size-tuning.md:169]: https://github.com/openshift/enhancements/pull/1549#discussion_r1465215686

Having a fully configurable limit lets customers better fit the limit to their own available memory.
The actual memory use will also be anything up to the limit (with some overhead for other things like the WAL file).

As long as the max limit is safe to use, discretizing the allowed limits in between (e.g in steps 16 GB, 32 GB, 64GB) provides no benefit in terms of performance or safety of the etcd cluster. Maybe a simpler API with enums but it trades off better utilization of the available resources.

File: enhancements/etcd/etcd-size-tuning.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+Customers have asked for the ability to increase the quota for the backend database to faciliate larger clusters.
+Specifically, in our build clusters, the number of namespaces and CRDs (and the size of the CRDs) have been increasing which drastically increase the amount of data stored in etcd.
+In 4.14, the perf team is now unable to run their benchmark suite.
+We want to be able to allow for this while minimizing the risk of them setting values that cause issues.
+
+### User Stories
+
+* As an administrator, I want to change the etcd backend size to decrease the risk of quota reached errors in very large clusters.
+* As an Openshift support, I want to walk a customer through changing the etcd size in a minimal number of steps.
+* As an etcd engineer, I want to easily add and test new sizes for performance impacts on both large and small clusters.
```
Comment (dofinn) [inline enhancements/etcd/etcd-size-tuning.md:39]: As an SRE, I want to safely scale up the size of etcd without impacting the availability of dependent components.

File: enhancements/etcd/etcd-size-tuning.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+### Goals
+
+* Add a configuration option using resource.Quantity to allow for human readable values; i.e. 16GiB.
+* Remove the parameters from the podspec rendering and replace with the value.
+* Add an API to allow admins to change the value.
+* Add blocker to disallow shrinking the size once set to a larger value.
```
Comment (dofinn) [inline enhancements/etcd/etcd-size-tuning.md:46]: Prevent resizing of etcd into an un-safe operational condition where its new size impacts control plane components due to available resources.

File: enhancements/etcd/etcd-size-tuning.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+  - With validation disallow a decrease in value.
+
+### Implementation Details/Notes/Constraints [optional]
+
+None
+
+### Risks and Mitigations
+
+* We will need to test the larger size to ensure that everything works as expected; especially defragmentation.
+* Defragmentation, and its effects, will need to be tested as the API server will be unresponsive during this time; we will need to understand the general magnitude of defrag time and documentation those expectations.
```
Comment (dofinn) [inline enhancements/etcd/etcd-size-tuning.md:89]: This will impact API Server SLOs. Having a bigger size, greater defragmentation == less potential availability for the apiserver. SLA implications for managed services.

File: enhancements/etcd/etcd-size-tuning.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+* There will be a required etcd rollout when changing values, this happens automatically by the Cluster Etcd Operator to re-render the etcd podspec.
+
+## Design Details
+
+### Open Questions [optional]
+
+For each of these, we would need to compare against the current default of 8GiB. Increasing the quota does not necessarily increase the following, but at the extreme case (e.g. increase to 16GiB and using nearly that much) we expect the following to perform differently.
+* How much memory do we require for a given database size?
+* How long does it take to sync a snapshot of a given size to another peer; for reboot/recovery?
```
Comment (dofinn) [inline enhancements/etcd/etcd-size-tuning.md:106]: Should consider reframing this to "what is an acceptable amount of downtime for the apiserver during this process"

File: enhancements/etcd/etcd-size-tuning.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
```
Comment (dofinn) [inline enhancements/etcd/etcd-size-tuning.md:154]: Should this resize be preventing during a CPMS scaling event?

**Review by dofinn — COMMENTED** [review]
Left some notes from SRE perspective :)

File: enhancements/etcd/etcd-size-tuning.md (hunk 12 above)
Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:154]: not sure how setting an env variable via static pods would create any downtime or should be prevented for CPMS updates, care to elaborate?

File: enhancements/etcd/etcd-size-tuning.md (hunk 10 above)
Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:89]: managed services do not defrag via the etcd operator. I also just want to leave a quick note that etcd still has two functional other members to reach while CEO defragments one, so I doubt this causes any availability issues.

@dusk125 can you update the scope/non-goals to not apply to managed services please? And maybe defuse this section, the etcd member under defrag will not be able to serve any writes, that doesn't mean the apiserver is unavailable.

File: enhancements/etcd/etcd-size-tuning.md (hunk 6 above)
Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:161]: @dhellmann the setting in question is merely a quota in an env variable. Increasing it has no immediate effect on resource utilization. Monitoring a static pod rollout would be just inspecting `oc get etcd/cluster` as usual.

@dusk125 can we also clarify this somewhere? it seems most people understand that after setting it you're suddenly down 50 gigs of disk space and the whole control plane is at an availability risk.

File: enhancements/etcd/etcd-size-tuning.md (hunk 5 above)
Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:156]: etcd does not detect this, but we can add an alert. Not sure we have disk space metrics in prometheus though. 

Also imagine the case where one CP node has less disk available (eg due to bigger log files on the same partition). In an ideal world we would need to gather the metrics for all CP node disks utilization and decide.

Ideally would be a quota, or even a dedicated disk drive / partition that we can rely on - but that's sadly not the happy path installation experience in OCP as of today.

> db in usage size grows larger than the available memory

That's okay and not a problem. It's a memory mapped file, not an in-memory-file :)

Given the perf test results from [1] I would also assume apiserver going OOM long before etcd even gets there. Where we both have CP memory alerts on. 

[1] https://docs.google.com/presentation/d/1a_LiZzWgCepx0ywrO4nmNqMEpdtuU0j9CcrSViX6Vjs/edit#slide=id.g2a7e4538e11_0_64

File: enhancements/etcd/etcd-size-tuning.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:80]: > a name like backendQuotaGiB and be an integer.

indeed. I think that's a good compromise between readability and ease of validation.

The `resource.Quantity` would give the option for fractional values like `5.6GiB` or `5600MiB`, but I don't think it's worth going below the gigabyte sizes here. ushift as an SKU would be a concern, but it's not running the CEO to manage etcd.

[the remark this answers is not in the document]
Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:80]: In case anyone ever wonders how such a validation would look like and before it gets lost in our internal slack:

> +kubebuilder:validation:XValidation:rule="self.CmpInt64(oldSelf)>0",message="can't decrease database size"

File: enhancements/etcd/etcd-size-tuning.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+### Drawbacks
+
+* There will be a required etcd rollout when changing values, this happens automatically by the Cluster Etcd Operator to re-render the etcd podspec.
+
+## Design Details
+
+### Open Questions [optional]
+
+For each of these, we would need to compare against the current default of 8GiB. Increasing the quota does not necessarily increase the following, but at the extreme case (e.g. increase to 16GiB and using nearly that much) we expect the following to perform differently.
+* How much memory do we require for a given database size?
```
Comment (soltysh) [inline enhancements/etcd/etcd-size-tuning.md:105]: I was wondering should we add a validation or at least a warning explaining which quota size requires how much memory? We should strive to protect users from shooting in the foot, but using 64GB quota on small machines.

File: enhancements/etcd/etcd-size-tuning.md (hunk 5 above)
Comment (soltysh) [reply enhancements/etcd/etcd-size-tuning.md:156]: See my other comment about ram, which should also be cross-checked with storage.

File: enhancements/etcd/etcd-size-tuning.md (hunk 7 above)
Comment (dhellmann) [reply enhancements/etcd/etcd-size-tuning.md:169]: It would be good to capture that in the doc.

**Comment by dhellmann** [comment]
#1555 is changing the enhancement template in a way that will cause the header check in the linter job to fail for existing PRs. If this PR is merged within the development period for 4.16 you may override the linter if the only failures are caused by issues with the headers (please make sure the markdown formatting is correct). If this PR is not merged before 4.16 development closes, please update the enhancement to conform to the new template.

File: enhancements/etcd/etcd-size-tuning.md (hunk 10 above)
Comment (dofinn) [reply enhancements/etcd/etcd-size-tuning.md:89]: etcd API would be where my assumption was wrong

File: enhancements/etcd/etcd-size-tuning.md (hunk 12 above)
Comment (dofinn) [reply enhancements/etcd/etcd-size-tuning.md:154]: assuming there is timeouts on the resize operation, if attempting to perform this during a control plane scaling event may cause it to fail? Just wondering about what potential failure modes may become apparent.

Comment (dofinn) [reply enhancements/etcd/etcd-size-tuning.md:154]: Thank you for the education

File: enhancements/etcd/etcd-size-tuning.md [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
+- Add BackendQuotaGiB of type int64
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
+  - With an upper limit of '64GiB'.
+
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
```
Comment (Elbehery) [inline enhancements/etcd/etcd-size-tuning.md:88]:
```suggestion
  // BackendQuotaGiB sets the etcd backend storage size limit in giga bytes. Defaults to 8GiB.
```

File: enhancements/etcd/etcd-size-tuning.md [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:default=8
+  // +kubebuilder:validation:Maximum=64
+  // +kubebuilder:validation:XValidation:rule="self.CmpInt64(oldSelf)>0",message="can't decrease database size"
```
Comment (Elbehery) [inline enhancements/etcd/etcd-size-tuning.md:92]:
```suggestion
  // +kubebuilder:validation:XValidation:rule="self >= oldSelf",message="can't decrease database size"
```

Comment (Elbehery) [reply enhancements/etcd/etcd-size-tuning.md:92]: during running the e2e test against these changes, `CmpInt64` failed.

Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:92]: int64 does not implement CmpInt64

File: enhancements/etcd/etcd-size-tuning.md (hunk 14 above)
Comment (tjungblu) [reply enhancements/etcd/etcd-size-tuning.md:88]: Gibibytes is correct, because we translate into base-2, don't we?

File: enhancements/etcd/etcd-size-tuning.md [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+## Summary
+
+This enhancement would replace the hardcoded value for the etcd backend database quota with a human readable, configurable value; 8GiB. This value does not immediately reserve disk space for etcd; it is an upper limit to the size of the etcd database on disk.
+This enhancement only covers the mvp for a tech preview release of this new feature; a future enhancement will be necessary.
+
+## Motivation
+
+Customers have asked for the ability to increase the quota for the backend database to faciliate larger clusters.
+Specifically, in our build clusters, the number of namespaces and CRDs (and the size of the CRDs) have been increasing which drastically increase the amount of data stored in etcd.
```
[the remark this answers is not in the document]
Comment (Elbehery) [reply enhancements/etcd/etcd-size-tuning.md:31]: cc @williamcaban

**Comment by dusk125** [comment]
Update for those following along at home:

Perfscale ran some tests and found that 32GiB seems to be a maximum; between 34GiB and 40GiB began to cause problems. The PR has been updated to have 32GiB as the enforced maximum.

There is a known issue in the API Server that can cause it to fall into a boot loop on restart due to watchers all trying to request large amounts of data all at the same time. This issue could be more likely to occur by larger etcd databases. The API Server team is working on addressing this issue, and this issue should be a GA blocker for this feature: the PR has been updated to include this as a requirement prior to GA.

File: enhancements/etcd/etcd-size-tuning.md [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+  - With an upper limit of '32GiB'.
+
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:default=8
+  // +kubebuilder:validation:Maximum=32
+  // +kubebuilder:validation:XValidation:rule="self>=oldSelf",message="can't decrease database size"
+  // +openshift:enable:FeatureGates=EtcdBackendQuota
```
[the remark this answers is not in the document]
Comment (Elbehery) [reply enhancements/etcd/etcd-size-tuning.md:93]: @tjungblu wdyt about the naming (i.e. FeatureGate=EtcdBackendQuota) 

or shall we keep it `BackendQuotaGiB` as the field name ? 

cc @soltysh @deads2k

File: enhancements/etcd/etcd-size-tuning.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (soltysh) [reply enhancements/etcd/etcd-size-tuning.md:31]: This is about all kind of build/CI/CD clusters, for example our customers using tekton ci run into similar problems. I'd rephrase in general CI/CD/pipeline clusters. Our build clusters are the closest to us, but not the only ones requesting this topic.

File: enhancements/etcd/etcd-size-tuning.md [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+* Handle consuming value changes without an etcd rollout.
+
+## Proposal
+
+Allow the customer/admin access to set the backend quota directly through a human readable value via the etcd.operator.openshift.io/v1 techpreview field. The only restriction is that they are disallowed from setting the value lower than the current.
+
+Increasing this value will not have etcd immediately reserve this space on disk, it just allows etcd to grow up to that value if etcd grows under the usage of the cluster.
+
+The default value or unset value would be the current, hardcoded backend quota: 8GiB. This would allow for upgrades where this configuration value wouldn't nesscessarily be set.
+An initial upper limit of 32GiB will be enforced so as to limit the impact of some etcd operations, like restore, transporting snapshots, etc. 32GiB is an arbitrary value and work with the Perf-Scale team will need to be done to validate this value and potentially increase the upper limit; finding a reasonable and safe upper bound will need to be done before this feature is made GA.
```
Comment (soltysh) [inline enhancements/etcd/etcd-size-tuning.md:58]: Based on the work in https://issues.redhat.com/browse/PERFSCALE-2881 this is not arbitrary number anymore. But rather a proven value, where our perf team tested up to 40G being the point where cluster was unstable and unrecoverable.

File: enhancements/etcd/etcd-size-tuning.md [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+None
+
+### Dev Preview -> Tech Preview
+
+N/A
+
+### Tech Preview -> GA
+
+- Known issue in API Server
+  - When the API server restarts, it's currently possible for all watchers to be reset at the same time and request a large amount of data from etcd, leading to a crash-loop in the API server.
```
Comment (soltysh) [inline enhancements/etcd/etcd-size-tuning.md:168]: This was also brought up in the slack, we want to ensure we have a ready procedure for cases when the cluster enters the crash-loop with apiserver. Even after apiserver fixes both bits:
- apistreaming 
- cbor for CRDs
The team will have to create condition to get to the point where we are still able to create that crash-loop and write down a document how to recover from such a state. 

Also I'd like to see both of the above apiserver dependencies (ideally with links to upstream KEPs) being explicitly linked in this proposal, so it's easier to track the current status of those.

File: enhancements/etcd/etcd-size-tuning.md [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+- Add BackendQuotaGiB of type int64
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
+  - With an upper limit of '32GiB'.
+
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:default=8
```
Comment (Elbehery) [inline enhancements/etcd/etcd-size-tuning.md:90]: I have a case to argue about the default value tag 

- you can find in the default [test](blob:https://prow.ci.openshift.org/009d7cee-20a8-4c1f-9ba4-be4f99a8314c), that expects the `EtcdSpec` to be empty, it expects  `"backendQuotaGiB": <int64>8` 
- this is due to the fact that the field is not a pointer
- since the field is `int64` it should defaults to `0`
- however, because of this tag, it defaults to `8`
- I propose to replace the default tag with ` +kubebuilder:validation:Minimum=8` to avoid these issues
- However, as I have initially proposed a `pointer` field would have been more suitable to avoid these situations


Possible solutions 

- use Pointer field
- use `+kubebuilder:validation:Minimum=8`

cc @tjungblu @dusk125 @hasbro17 @deads2k @soltysh

Comment (Elbehery) [reply enhancements/etcd/etcd-size-tuning.md:90]: see https://github.com/openshift/api/pull/1736#issuecomment-2069623428 for details

Comment (Elbehery) [reply enhancements/etcd/etcd-size-tuning.md:90]: first time that the integration passed 🟢  [link](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_api/1736/pull-ci-openshift-api-master-integration/1782424939345743872)

I used these tags, can u please update the enhancement accordingly 


```
        // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
	// +kubebuilder:validation:Optional
	// +kubebuilder:validation:Minimum=8
	// +kubebuilder:validation:Maximum=32
	// +kubebuilder:validation:XValidation:rule="self>=oldSelf",message="can't decrease database size"
	// +openshift:enable:FeatureGate=EtcdBackendQuota
	// +optional
```

File: enhancements/etcd/etcd-size-tuning.md [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=8
+	// +kubebuilder:validation:Maximum=32
+	// +kubebuilder:validation:XValidation:rule="self>=oldSelf",message="can't decrease database size"
+	// +openshift:enable:FeatureGate=EtcdBackendQuota
+	// +optional
+  BackendQuotaGiB int64 `json:"backendQuotaGiB"`
```
Comment (JoelSpeed) [inline enhancements/etcd/etcd-size-tuning.md:95]: Given discussion on the API, this value will always be between 8 and 32. An `int32` would be sufficient here (only `int32` and `int64` are supported kinds of ints for Kube APIs)

Comment (Elbehery) [reply enhancements/etcd/etcd-size-tuning.md:95]: `lgtm` 👍🏽  

I will update the API here 

@dusk125  would you kindly update the enhancement accordingly ?

Comment (dusk125) [reply enhancements/etcd/etcd-size-tuning.md:95]: I'm good with that, I'll update

**Comment by Elbehery** [comment]
@hasbro17 @tjungblu @soltysh @deads2k @JoelSpeed 

Any more reviews, or shall we merge ?

**Comment by hasbro17** [comment]
/approve

The enhancement looks good to me for a tech-preview implementation (sans any more API specifics which we can loop back in from https://github.com/openshift/api/pull/1736).

Seems like we've explicitly mentioned the GA requirements for more performance testing, validating the API server's ability to handle large requests at the higher limit, investigating the OOM cycle scenario and documenting the recovery procedure for that.

I'll defer the LGTM to @soltysh or @deads2k in case there are more scenarios we need to callout here.

File: enhancements/etcd/etcd-size-tuning.md [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+A downside is that with the profiles, discussed in the Alternatives section, there are a discrete, small, number of permutations to test, giving more confidence of the exact effects a given profile has on performance/memory requirements.
+This is a minor consideration as setting the quota to a arbitrarily higher value does not mean that etcd will take up that much space; it will depend on the workload.
+
+### Workflow Description
+
+**cluster administrator** is a human user responsible deploying and maintaining the cluster.
+
+1. The cluster administrator decides to change the etcd backend quota from the default (8GiB) to a larger value; e.g. 20GiB.
+2. They set the new value via the `etcd.operator.openshift.io/v1` API.
+3. An etcd redeployment will be automatically run which restarts the etcd pods which consume the new value.
```
Comment (soltysh) [inline enhancements/etcd/etcd-size-tuning.md:70]:
```suggestion
3. An etcd redeployment will be automatically run which restarts the etcd pods with the new quota set.
```

File: enhancements/etcd/etcd-size-tuning.md [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=8
+	// +kubebuilder:validation:Maximum=32
+	// +kubebuilder:validation:XValidation:rule="self>=oldSelf",message="can't decrease database size"
+	// +openshift:enable:FeatureGate=EtcdBackendQuota
+	// +optional
+  BackendQuotaGiB int32 `json:"backendQuotaGiB"`
```
Comment (soltysh) [inline enhancements/etcd/etcd-size-tuning.md:95]: This will need to be updated once https://github.com/openshift/api/pull/1736 merges, but that's not a blocker.

File: enhancements/etcd/etcd-size-tuning.md [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+* Add a configuration option using resource.Quantity to allow for human readable values; i.e. 16GiB.
+* Remove the parameters from the podspec rendering and replace with the value.
+
+### Risks and Mitigations
+
+* We will need to test larger sizes to ensure that everything works as expected.
+* For a larger database, defragmentation of etcd will take longer which will increase the amount of time that etcd is unavaiable for writes.
+  Etcd will still be available for reads during defragmentation, so there should be little impact on the apiserver's availability.
+* Downgrades may be impacted if:
+  1. The downgrade is to a version without this feature.
```
Comment (soltysh) [inline enhancements/etcd/etcd-size-tuning.md:127]: Probably worth considering a way to resolve this problem in the future, not a blocker atm.

Comment (soltysh) [reply enhancements/etcd/etcd-size-tuning.md:127]: One potential approach would be to add a check that this is not set in a version prior to when this feature goes GA. To prevent that kind of downgrade, or even block it.

**Review by soltysh — COMMENTED** [review]
minor comments, overall lgtm

File: enhancements/etcd/etcd-size-tuning.md [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+#### Single-node Deployments or MicroShift
+
+This feature should not change the resource consumption of a single-node OpenShift deployment. If an admin decides to increase the quota, the same caveats, from normal OpenShift, about disk space will apply - etcd will not immediately reserve the quota-size, but could grow up to the quota and exhaust disk space in the process.
+
+This enhancement will not directly impact MicroShift as the api will be consume via the Cluster Etcd Operator.
+[Also, changing this value is already supported in the MicroShift configuration](https://github.com/openshift/microshift/blob/main/pkg/config/etcd.go#L18).
+
+### Implementation Details/Notes/Constraints
+
+* Add a configuration option using resource.Quantity to allow for human readable values; i.e. 16GiB.
```
Comment (JoelSpeed) [inline enhancements/etcd/etcd-size-tuning.md:118]: Is this still accurate?

File: enhancements/etcd/etcd-size-tuning.md [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+
+### Dev Preview -> Tech Preview
+
+N/A
+
+### Tech Preview -> GA
+
+- Known issue in API Server
+  - When the API server restarts, it's currently possible for all watchers to be reset at the same time and request a large amount of data from etcd, leading to a crash-loop in the API server.
+  - [https://github.com/kubernetes/enhancements/pull/4453](API Streaming)
```
Comment (soltysh) [inline enhancements/etcd/etcd-size-tuning.md:173]: I'd suggest pointing to the enhancement tracking issue which is https://github.com/kubernetes/enhancements/issues/3157

**Review by soltysh — COMMENTED** [review]
Minor nits that would be nice to get resolved and it's good to go.

**Review by soltysh — APPROVED** [review]
/lgtm
